### ISSUE TITLE
Indenting cuts off element's labels

### ISSUE THREAD

**Issue by darkgnostic** [ocornut/imgui#3056]
Imgui version: v1.76 WIP

When I use Indent & Unindent, the elements don't fit the screen width anymore, and the labels go out of screen.
On the following image labels **Hover EID, Hover UID** are just cut off.

Is this normal behavior? Or how can I achieve that the element's length don't go out of window width?

I tried to add _ImGuiWindowFlags_AlwaysAutoResize_ as a flag, but that didn't helped.

![Capture](https://user-images.githubusercontent.com/4980221/76324591-8c0ec600-62e6-11ea-9314-dd2a88070ac0.PNG)

```
const auto windows_flag_ = ImGuiWindowFlags_NoCollapse | ImGuiWindowFlags_NoBackground | ImGuiWindowFlags_NoTitleBar | ImGuiWindowFlags_NoResize | ImGuiWindowFlags_NoMove;
ImGui::Begin("Example Bug", 0, windows_flag_ );
ImGui::Indent();
ImGui::SliderFloat("Zoom", &zoom_, 0.1f, 6.0f);
ImGui::InputInt("Hover EID", &hover_eid_);
ImGui::Unindent();
ImGui::End();
```

**Comment by ocornut** [comment]
Hello,

I don't think the cut off has anything to do with indenting, there's no mechanism to prevent that in the current item layout idiom.

By default for normal-resizable window the default item width (without label) is `0.65f * window_width`, you can use `PushItemWidth()` to alter that value. See `Demo>Layout>Widget Width` for details.

If you use `SetNextWindowSize()` that's overriding the `ImGuiWindowFlags_AlwaysAutoResize` flag, so I would suggest you just remove the `SetNextWindowSize()` call.
Note that if you enable `ImGuiWindowFlags_AlwaysAutoResize` the default value for PushItemWidth() cannot be `0.65f * window_width` as it would create a feedback loop, instead it is FontSize*16 by default so you'll even more want to use `PushItemWidth()` to alter it.

TL;DR; You are hitting on a surprisingly old and unpolished part of the library (this is essentially #267 and #395). I have some ideas for it but it somehow managed to not be fixed for a long time. Data-driven property viewers tend to lay label separately from widget contents for that reason.

**Comment by darkgnostic** [comment]
Thanks for your reply. I need to work on this a bit to see what is the best solution.

**Comment by ocornut** [comment]
I think the two steps we want to head toward to solve that more automatically so you'd have nothing to do would be:

- Allow to associate that width to 1 or 2 manually resizable column (those sizes would need to persist). 
- Expose various combinations of "local layout" within standard widget so it becomes easier to have "Label, new line, Framed widget" or "Label, Framed Widget" the later is more easily possible now that we have soft text clipping with ellipsis.

**Comment by darkgnostic** [comment]
Thank you for your help.

Actually I think I have found another possible solution for it. If I compare window->ContentSize to window->Size and adjust item width with PushItemWidth, then items align nicely .

I just need to take into the account ScrollBar. How can I check if scroll bar is visible or not?

And is this idea also viable? Currently result look pretty ok:

![Capture](https://user-images.githubusercontent.com/4980221/76344876-95a62700-6302-11ea-935d-43186e691c93.PNG)

**Comment by ocornut** [comment]
> And is this idea also viable? Currently result look pretty ok:

I don't know, as I don't understand your description and it seems ambiguous to me.

**Comment by darkgnostic** [comment]
If I understand correctly then window->ContentSize is size of current window with clipped region. If I check just beforee ImGui::End() if window->ContenSize.x is greater than window->Size.x, then I know that some clipping occurred.

Here is a dirty example that seems to do the job (just a proof of concept):
```
IMGUI_API bool ImGui::IsWindowClipped()
{
    ImGuiContext& g = *GImGui;
    ImGuiWindow* window = g.CurrentWindow;
    const ImGuiStyle& style = g.Style;

    const float realWidth = window->ContentSize.x;
    const float innerWindowWidth = window->Size.x-window->WindowPadding.x-style.ScrollbarSize;
    return realWidth > innerWindowWidth;
}
....
   static float currentAspect = 0.65f;
   const float fixWindowSize = 200.0f;
   ImGui::SetNextWindowSize(ImVec2(fixWindowSize ,0));
   ImGui::Begin("Properties");
   ImGui::PushItemWidth( fixWindowSize * currentAspect );

   // draw some controls

    if( ImGui::IsWindowClipped() ) {
        currentAspect -= 0.01f;
    }
    ImGui::End()
```

**Comment by darkgnostic** [comment]
I have managed to fix this issue, but with slightly modifying the imgui_internal.h

Here is the code:

in _imgui_internal.h_ we need one member variable
```
struct IMGUI_API ImGuiWindow {
....
float                   ItemWidthAspect;                    // Aspect of the item width if width of the window is fix size
}
```
then in _imgui.cpp_'s constructor we initialize it
```
ImGuiWindow::ImGuiWindow(ImGuiContext* context, const char* name) {
....
ItemWidthAspect = -1.0f;
}
```

These two functions can go inside your application:
```
IMGUI_API void ImGui::PushItemAutoWidth()
{
	ImGuiContext& g = *GImGui;
	ImGuiWindow* window = g.CurrentWindow;
	
	if( window->AutoFitFramesX == 0 ) {

		if( window->ItemWidthAspect < 0.0f ) {
			window->ItemWidthAspect = 0.65f;
		}

		ImGui::PushItemWidth( g.NextWindowData.SizeVal.x * window->ItemWidthAspect);
	}
}

IMGUI_API void ImGui::CalcItemAutoWidth()
{
	ImGuiContext& g = *GImGui;
	ImGuiWindow* window = g.CurrentWindow;
	if( window->AutoFitFramesX == 0 ) {

		const ImGuiStyle& style = g.Style;

		const float realWidth = window->ContentSize.x;
		const float innerWindowWidth = window->Size.x-window->WindowPadding.x-(window->ScrollbarY ? style.ScrollbarSize : 0);

		if( realWidth != 0.0f )
			window->ItemWidthAspect += realWidth > innerWindowWidth ? -0.01f : ( realWidth > innerWindowWidth-10 ? 0 : 0.01f);
	}
}
````

So how it works? 

If you have your window for example made fixed sized:
```
ImGui::SetNextWindowSize(ImVec2(200,0)));
```

then by adding _ImGui::PushItemAuto_ width just after the _ImGui::Begin_, and _ImGui::CalcItemAutoWidth_ just before the _ImGui::End_, all items in that window will auto fit the content.

```
ImGui::Begin("Properties", 0, windows_flag_);
ImGui::PushItemAutoWidth();

....

ImGui::CalcItemAutoWidth();
ImGui::End();
```

Of course, you can simply add these inside the Begin/End, they will have effect on content only if the window width is explicitly set, but I didn't wanted to touch Imgui core code more than adding new member variable.
